### ISSUE TITLE
Group label not associated with its related controls (missing fieldset)

### ISSUE THREAD

**Issue by vikrampm1** [woothemes/woocommerce#58343]
### Issue Description:

This element has a visible label that provides context for a group of buttons or inputs. However, the visible label is not programmatically linked to the controls in the group, such as through a fieldset and legend, which means screen reader users do not receive the context the group label offers.

<img width="1156" alt="Image" src="https://github.com/user-attachments/assets/9ba5d972-a107-48dc-aaa3-afed1bd046fd" />

### Component or Page: 

Filters (v.9.9)  

### Name affected element:

All groups of checkboxes and chips

### Relevant code snippets:

```
Code example: 
<div data-block-name=""woocommerce/product-filter-rating"" data-show-counts=""true"" class=""wp-block-woocommerce-product-filter-rating"" data-wp-interactive=""woocommerce/product-filters"" data-wp-key=""woocommerce/product-filter-rating2"" data-wp-context=""{&quot;activeLabelTemplate&quot;:&quot;Rating: {{label}}&quot;,&quot;filterType&quot;:&quot;rating&quot;}"">
<h3 class=""wp-block-heading"" style=""margin-top:0;margin-bottom:0.625rem"">Rating</h3>
		<div data-block-name=""woocommerce/product-filter-checkbox-list"" class=""wp-block-woocommerce-product-filter-checkbox-list wc-block-product-filter-checkbox-list wp-block-woocommerce-product-filter-checkbox-list"" data-wp-interactive=""woocommerce/product-filters"" data-wp-key=""woocommerce/product-filter-checkbox-list4"" data-wp-context=""{}"">
			<ul class=""wc-block-product-filter-checkbox-list__list"" aria-label=""Filter Options"">
														<li data-wp-key=""rating-5"" class=""wc-block-product-filter-checkbox-list__item"">
						<label class=""wc-block-product-filter-checkbox-list__label"" for=""rating-5"">
							<span class=""wc-block-product-filter-checkbox-list__input-wrapper"">
								<input id=""rating-5"" class=""wc-block-product-filter-checkbox-list__input"" type=""checkbox"" aria-label=""Rated 5 out of 5"" data-wp-on--change=""actions.toggleFilter"" value=""5"" data-wp-bind--checked=""state.isFilterSelected"" data-wp-context=""{&quot;item&quot;:{&quot;label&quot;:&quot;\t\t\u003Cdiv class=\u0022wc-block-components-product-rating\u0022\u003E\n\t\t\t\u003Cdiv class=\u0022wc-block-components-product-rating__stars\u0022 role=\u0022img\u0022 aria-label=\u0022Rated 5 out of 5\u0022\u003E\n\t\t\t\t\u003Cspan style=\u0022width: 100%\u0022 aria-hidden=\u0022true\u0022\u003E\u003C\/span\u003E\n\t\t\t\u003C\/div\u003E\n\t\t\u003C\/div\u003E\n\t\t&quot;,&quot;ariaLabel&quot;:&quot;Rated 5 out of 5&quot;,&quot;value&quot;:&quot;5&quot;,&quot;selected&quot;:false,&quot;count&quot;:1,&quot;type&quot;:&quot;rating&quot;}}"">
								<svg class=""wc-block-product-filter-checkbox-list__mark"" viewBox=""0 0 10 8"" fill=""none"" xmlns=""http://www.w3.org/2000/svg"">
									<path d=""M9.25 1.19922L3.75 6.69922L1 3.94922"" stroke=""currentColor"" stroke-width=""1.5"" stroke-linecap=""round"" stroke-linejoin=""round""></path>
								</svg>
							</span>
							<span class=""wc-block-product-filter-checkbox-list__text-wrapper"">
								<span class=""wc-block-product-filter-checkbox-list__text"">
											<div class=""wc-block-components-product-rating"">
			<div class=""wc-block-components-product-rating__stars"" role=""img"" aria-label=""Rated 5 out of 5"">
				<span style=""width: 100%"" aria-hidden=""true""></span>
			</div>
		</div>
										</span>
																	<span class=""wc-block-product-filter-checkbox-list__count"">
										(1)
									</span>
															</span>
						</label>
					</li>
														<li data-wp-key=""rating-4"" class=""wc-block-product-filter-checkbox-list__item"">
						<label class=""wc-block-product-filter-checkbox-list__label"" for=""rating-4"">
							<span class=""wc-block-product-filter-checkbox-list__input-wrapper"">
								<input id=""rating-4"" class=""wc-block-product-filter-checkbox-list__input"" type=""checkbox"" aria-label=""Rated 4 out of 5"" data-wp-on--change=""actions.toggleFilter"" value=""4"" data-wp-bind--checked=""state.isFilterSelected"" data-wp-context=""{&quot;item&quot;:{&quot;label&quot;:&quot;\t\t\u003Cdiv class=\u0022wc-block-components-product-rating\u0022\u003E\n\t\t\t\u003Cdiv class=\u0022wc-block-components-product-rating__stars\u0022 role=\u0022img\u0022 aria-label=\u0022Rated 4 out of 5\u0022\u003E\n\t\t\t\t\u003Cspan style=\u0022width: 80%\u0022 aria-hidden=\u0022true\u0022\u003E\u003C\/span\u003E\n\t\t\t\u003C\/div\u003E\n\t\t\u003C\/div\u003E\n\t\t&quot;,&quot;ariaLabel&quot;:&quot;Rated 4 out of 5&quot;,&quot;value&quot;:&quot;4&quot;,&quot;selected&quot;:false,&quot;count&quot;:1,&quot;type&quot;:&quot;rating&quot;}}"">
								<svg class=""wc-block-product-filter-checkbox-list__mark"" viewBox=""0 0 10 8"" fill=""none"" xmlns=""http://www.w3.org/2000/svg"">
									<path d=""M9.25 1.19922L3.75 6.69922L1 3.94922"" stroke=""currentColor"" stroke-width=""1.5"" stroke-linecap=""round"" stroke-linejoin=""round""></path>
								</svg>
							</span>
							<span class=""wc-block-product-filter-checkbox-list__text-wrapper"">
								<span class=""wc-block-product-filter-checkbox-list__text"">
											<div class=""wc-block-components-product-rating"">
			<div class=""wc-block-components-product-rating__stars"" role=""img"" aria-label=""Rated 4 out of 5"">
				<span style=""width: 80%"" aria-hidden=""true""></span>
			</div>
		</div>
										</span>
																	<span class=""wc-block-product-filter-checkbox-list__count"">
										(1)
									</span>
															</span>
						</label>
					</li>
														<li data-wp-key=""rating-2"" class=""wc-block-product-filter-checkbox-list__item"">
						<label class=""wc-block-product-filter-checkbox-list__label"" for=""rating-2"">
							<span class=""wc-block-product-filter-checkbox-list__input-wrapper"">
								<input id=""rating-2"" class=""wc-block-product-filter-checkbox-list__input"" type=""checkbox"" aria-label=""Rated 2 out of 5"" data-wp-on--change=""actions.toggleFilter"" value=""2"" data-wp-bind--checked=""state.isFilterSelected"" data-wp-context=""{&quot;item&quot;:{&quot;label&quot;:&quot;\t\t\u003Cdiv class=\u0022wc-block-components-product-rating\u0022\u003E\n\t\t\t\u003Cdiv class=\u0022wc-block-components-product-rating__stars\u0022 role=\u0022img\u0022 aria-label=\u0022Rated 2 out of 5\u0022\u003E\n\t\t\t\t\u003Cspan style=\u0022width: 40%\u0022 aria-hidden=\u0022true\u0022\u003E\u003C\/span\u003E\n\t\t\t\u003C\/div\u003E\n\t\t\u003C\/div\u003E\n\t\t&quot;,&quot;ariaLabel&quot;:&quot;Rated 2 out of 5&quot;,&quot;value&quot;:&quot;2&quot;,&quot;selected&quot;:false,&quot;count&quot;:2,&quot;type&quot;:&quot;rating&quot;}}"">
								<svg class=""wc-block-product-filter-checkbox-list__mark"" viewBox=""0 0 10 8"" fill=""none"" xmlns=""http://www.w3.org/2000/svg"">
									<path d=""M9.25 1.19922L3.75 6.69922L1 3.94922"" stroke=""currentColor"" stroke-width=""1.5"" stroke-linecap=""round"" stroke-linejoin=""round""></path>
								</svg>
							</span>
							<span class=""wc-block-product-filter-checkbox-list__text-wrapper"">
								<span class=""wc-block-product-filter-checkbox-list__text"">
											<div class=""wc-block-components-product-rating"">
			<div class=""wc-block-components-product-rating__stars"" role=""img"" aria-label=""Rated 2 out of 5"">
				<span style=""width: 40%"" aria-hidden=""true""></span>
			</div>
		</div>
										</span>
																	<span class=""wc-block-product-filter-checkbox-list__count"">
										(2)
									</span>
															</span>
						</label>
					</li>
							</ul>
					</div>
		</div>
```

### Recommended fix:

Wrap the entire component in a fieldset tag and use a legend tag to connect the group label with the sub elements. For more information refer to: https://developer.mozilla.org/en-US/docs/Web/HTML/Element/fieldset

**Comment by dinhtungdu** [comment]
> Wrap the entire component in a fieldset tag and use a legend tag to connect the group label with the sub elements. For more information refer to: [developer.mozilla.org/en-US/docs/Web/HTML/Element/fieldset](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/fieldset)

@vikrampm1 @amberhinds @majomaldonado is there any alternative to fix this issue? Can we fix it using `aria-label` or `aria-labeledby` like this:

```
<div class="wp-block-woocommerce-product-filter-rating" aria-label="Rating">
...
</div>
```
Or
```
<div class="wp-block-woocommerce-product-filter-rating" aria-labelledby="heading-id-0">
    <h3 class="wp-block-heading" id="heading-id-0">
    ...
</div>

**Comment by SteveJonesDev** [comment]
@dinhtungdu,

### Aria Label:
Using `aria-label` on a `<div>` that contains interactive child elements is not the appropriate way to convey a group label and break ARIA taxonomy. 

### Aria Labelled By:
`aria-labelledby` could technically work, but may override the accessible names of the individual checkboxes if not applied correctly. This would need to be applied on a container with an appropriate `role`, like `role="group"` or `role="radiogroup"`. This would needs some thorough screenreader testing as it's not the ideal implementation.

### Semantic HTML
The most standards-compliant fix is to use a `<fieldset>` and `<legend>` to group fields. This is the recommended method per both WCAG and ARIA best practices.

**Comment by SteveJonesDev** [comment]
@dinhtungdu,

Here is an example of a possible implementation using `aria-labelledby`. I'd just ensure any `aria-label` inside of that group still reads out to screenreaders.

https://www.w3.org/WAI/tutorials/forms/grouping/#associating-related-controls-with-wai-aria:~:text=in%20every%20label.-,Associating%20related%20controls%20with%20WAI%2DARIA,-WAI%2DARIA%20provides

**Comment by dinhtungdu** [comment]
Thank you @SteveJonesDev for the suggestions.

> The most standards-compliant fix is to use a `<fieldset>` and `<legend>` to group fields. This is the recommended method per both WCAG and ARIA best practices.

The reason I asked for alternative is that we can't use fieldset and legend here. We use the heading block for the title, so injecting attributes to the heading is more reliable and less problematic than changing its HTML structure.

> This would need to be applied on a container with an appropriate role, like role="group"

Right, thanks for the head up. I will try WAI-ARIA to see how it goes.

**Comment by dinhtungdu** [comment]
Opened https://github.com/woocommerce/woocommerce/pull/58561 to fix this issue. It would be awesome if one of you can take a look.

**Comment by amberhinds** [comment]
Can you use a screen reader only legend that's hard coded?

Amber Hinds

Founder & CEO
Pronouns: she/her

Phone: 512-942-5858
https://equalizedigital.com
[image: Equalize Digital - websites for everyone logo]
------------------------------

Equalize Digital believes in digital equality. Our mission is to make the
web more accessible for our team members, for our clients, *for everyone*.
If there is something in this email or an attachment that does not work for
you or your assistive technology, please let me know. I’ll do my best to
assist you and resolve issues. Read our full accessibility statement
<https://equalizedigital.com/about/accessibility-statement/>

**Comment by dinhtungdu** [comment]
> Can you use a screen reader only legend that's hard coded?

@amberhinds I'm not sure I got what you meant. Can you clarify, please?

**Comment by amberhinds** [comment]
@dinhtungdu, you can put a fieldset around the options and then have a legend that is visually hidden and only available to screen readers that doesn't pull from the header. Since you know the attribute that has been selected, you could just hard-code that in. Something like:
`<legend class="screen-reader-text">Star Rating</legend>`

**Comment by dinhtungdu** [comment]
@amberhinds Looks like it doesn't work for list elements. In https://github.com/woocommerce/woocommerce/commit/a93e1dda482ed8715964f0e17b1216d2f7fe8106, I implemented your suggestion but macOS Voice Over only announces the group label for Chips, not Checkbox List which uses `<ul>`. See https://cloudup.com/cgrsKM3FxIX

**Comment by dinhtungdu** [comment]
Should we stop using `<ul>` for this A11y reason or should we go with https://github.com/woocommerce/woocommerce/pull/58561? I think using `<ul>` is semantically correct for Checkbox List.

**Comment by amberhinds** [comment]
While they can be in a ul, normally checkboxes are not grouped in lists.
That could add extra noise, so I would probably remove it in favor of the
standard fieldset grouping. See
https://developer.mozilla.org/en-US/docs/Web/HTML/Reference/Elements/input/checkbox
and
https://developer.mozilla.org/en-US/docs/Web/HTML/Reference/Elements/fieldset


Amber Hinds

Founder & CEO
Pronouns: she/her

Phone: 512-942-5858
https://equalizedigital.com
[image: Equalize Digital - websites for everyone logo]
------------------------------

Equalize Digital believes in digital equality. Our mission is to make the
web more accessible for our team members, for our clients, *for everyone*.
If there is something in this email or an attachment that does not work for
you or your assistive technology, please let me know. I’ll do my best to
assist you and resolve issues. Read our full accessibility statement
<https://equalizedigital.com/about/accessibility-statement/>
